django/db/models/deletion.py

```diff
@@ -212,7 +212,16 @@ def collect(self, objs, source=None, nullable=False, collect_related=True,
                                  collect_related=False,
                                  reverse_dependency=True)
         if collect_related:
-            parents = model._meta.parents
+            if keep_parents:
+                # Recursively collect all parent models.
+                def _get_parents(model_parents):
+                    parents.update(model_parents)
+                    for parent in model_parents:
+                        _get_parents(parent._meta.parents)
+
+                parents = {}
+                _get_parents(model._meta.parents)
```

Comment:
Maybe there is a smarter way to recursively collect all parents? e.g. some hook that I missed :thinking:  (\cc @charettes ).

Reply:
@felixxm, @december1981 the `_meta.get_parent_list` method seems to be what you're after

https://github.com/django/django/blob/29601bca9b584b254ded0a8f59cf776194f1bd28/django/db/models/options.py#L584-L593

The result should be converted to a `set` though to avoid O(n) containment lookups. e.g.

```python
parents = set(model._meta.get_parent_list())
```

Reply:
@charettes Many thanks!